src/Bicep.Core/Registry/Oci/TagEncoder.cs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Modules;
+using System;
+using System.Diagnostics;
+using System.Text;
+
+namespace Bicep.Core.Registry.Oci
+{
+    /// <summary>
+    /// Encodes the casing of an OCI tag into a file name that can be used with case-insensitive file systems.
+    /// </summary>
+    public static class TagEncoder
+    {
+        // not sure there's a way to get the constant programmatically
+        private const int BitsInUnsignedLong = 64;
+
+        private const int BitsInHexDigit = 4;
+
+        // this should be equivalent to ceil(OciArtifactModuleReference.MaxTagLength / 64.0) but without floating point conversion
+        private const int MaskComponents = (OciArtifactModuleReference.MaxTagLength + BitsInUnsignedLong - 1) / BitsInUnsignedLong;
+
+        private const string HexFormat = "x";
+
+        // TODO: change to const in .net 6
+        private static readonly string UnsignedLongHexFormat = $"{HexFormat}{BitsInUnsignedLong / BitsInHexDigit}";
+
+        /// <summary>
+        /// Encodes the specified OCI tag into a file name that can be used with case-insensitive file systems.
+        /// </summary>
+        /// <param name="tag">The tag value. The tag should be validated to match the OCI spec before calling this function.</param>
+        public static string Encode(string tag)
+        {
+            if(tag.Length > OciArtifactModuleReference.MaxTagLength)
+            {
+                throw new ArgumentException($"The specified tag '{tag}' exceeds max length of {OciArtifactModuleReference.MaxTagLength}.");
+            }
+
+            var mask = new ulong[MaskComponents];
+            for (int i = 0; i < tag.Length; i++)
+            {
+                int maskIndex = i / BitsInUnsignedLong;
+                ulong maskValue = 1UL << (i % BitsInUnsignedLong);
+
```

Comment:
I think we can move the declarations to the if body below to avoid unnecessary calculation.

Reply:
Will do

Reply:
Done